modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/upgrade/v8_9_3/CommerceOrderUpgradeProcess.java

```diff
@@ -0,0 +1,134 @@
+import com.liferay.commerce.internal.order.status.ShippedCommerceOrderStatusImpl;
+import com.liferay.commerce.model.CommerceOrder;
+import com.liferay.commerce.order.status.CommerceOrderStatus;
+import com.liferay.commerce.order.status.CommerceOrderStatusRegistry;
+import com.liferay.commerce.service.CommerceOrderLocalService;
+import com.liferay.commerce.service.CommerceShipmentLocalService;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+/**
+ * @author Brian I. Kim
+ */
+public class CommerceOrderUpgradeProcess extends UpgradeProcess {
+
+	public CommerceOrderUpgradeProcess(
+		CommerceOrderLocalService commerceOrderLocalService,
+		CommerceShipmentLocalService commerceShipmentLocalService,
+		CommerceOrderStatusRegistry commerceOrderStatusRegistry) {
+
+		_commerceOrderLocalService = commerceOrderLocalService;
+		_commerceShipmentLocalService = commerceShipmentLocalService;
+		_commerceOrderStatusRegistry = commerceOrderStatusRegistry;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		_updateCommerceOrder();
+	}
+
+	private void _updateCommerceOrder() throws Exception {
+		try (PreparedStatement preparedStatement = connection.prepareStatement(
+				"select commerceOrderId, orderStatus from CommerceOrder " +
+					"where orderStatus = ? or orderStatus = ? ")) {
+
+			preparedStatement.setInt(
+				1, CommerceOrderConstants.ORDER_STATUS_PARTIALLY_SHIPPED);
+			preparedStatement.setInt(
+				2, CommerceOrderConstants.ORDER_STATUS_SHIPPED);
+
+			try (ResultSet resultSet = preparedStatement.executeQuery()) {
+				while (resultSet.next()) {
+					int orderStatus = resultSet.getInt(2);
+
+					if (orderStatus ==
+							CommerceOrderConstants.
+								ORDER_STATUS_PARTIALLY_SHIPPED) {
+
+						CommerceOrder commerceOrder =
+							_commerceOrderLocalService.getCommerceOrder(
+								resultSet.getLong(1));
+
+						_updateShippedCommerceOrderStatus(commerceOrder);
+						_updateCompletedCommerceOrderStatus(commerceOrder);
+					}
+					else if (orderStatus ==
+								CommerceOrderConstants.ORDER_STATUS_SHIPPED) {
+
+						_updateCompletedCommerceOrderStatus(
+							_commerceOrderLocalService.getCommerceOrder(
+								resultSet.getLong(1)));
+					}
+				}
+			}
+		}
+	}
+
+	private void _updateCompletedCommerceOrderStatus(
+			CommerceOrder commerceOrder)
+		throws PortalException {
+
+		CommerceOrderStatus completedCommerceOrderStatus =
+			_commerceOrderStatusRegistry.getCommerceOrderStatus(
+				CompletedCommerceOrderStatusImpl.KEY);
+
+		int[] commerceShipmentStatuses =
+			_commerceShipmentLocalService.
+				getCommerceShipmentStatusesByCommerceOrderId(
+					commerceOrder.getCommerceOrderId());
+
+		if (completedCommerceOrderStatus.isTransitionCriteriaMet(
+				commerceOrder) &&
+			(commerceShipmentStatuses.length == 1) &&
```

Comment:
With this one we are only checking 1 shipment per order, but order items from the same order can be shipped with different shipments, so we must check all the shipment statuses not just the first one commerceShipmentStatuses[0].

Reply:
@luca-pellizzon , the query checks for distinct shipmentstatuses. It will always return more than 1 if there are multiple types of shipmentstatuses in a single order.

Reply:
This is the relevant query:

```
	<sql id="com.liferay.commerce.service.persistence.CommerceShipmentFinder.findCommerceShipmentStatusesByCommerceOrderId">
		<![CDATA[
			SELECT
				DISTINCT CommerceShipment.status
			FROM
				CommerceShipment
			LEFT JOIN
				CommerceShipmentItem ON
					CommerceShipmentItem.commerceShipmentId = CommerceShipment.commerceShipmentId
			INNER JOIN
				CommerceOrderItem ON
					CommerceOrderItem.commerceOrderItemId = CommerceShipmentItem.commerceOrderItemId
			WHERE
				CommerceOrderItem.commerceOrderId = ?
		]]>
	</sql>
```

https://github.com/liferay-commerce/liferay-portal/blob/47122202b5b7d05b00237736e559449bacff9c76/modules/apps/commerce/commerce-service/src/main/resources/META-INF/custom-sql/default.xml#L172

Reply:
@luca-pellizzon , Let me know if you are satisfied with my explanation unless my understanding is incorrect.